Sidebar: Better search, keyboard shortcuts, and "recently viewed"

ghengeveld opened the pull request:
Issue: #10757

## What I did

This completely overhauls the sidebar. It replaces the Tree to be more performant and to add keyboard navigation, as well as expand/collapse all functionality. The Search has been replaced to show results in a list instead of filtering the tree, improving usability and performance. Furthermore I've added a "recently opened" list so allow quickly switching between recently viewed components. Additionally, this fixes a bug in composition where stories with the same path/id between refs would both get highlighted when one was selected.

Subtleties:
- Scroll into view for highlighted and selected nodes
- Allow collapsing a node even if a story is selected inside
- Double expand/collapse via keyboard does expand all / collapse all

Video: https://youtu.be/sbiYhmL2j2I

## How to test

I've added stories for every major component and state in the sidebar. Besides that, it reuses existing logic that's already covered by unit tests.

This is going to require significant QA. Things to consider:
- Navigating the tree with the mouse
- Navigating with keyboard, both with tab and arrow keys
- Expanding, collapsing nodes + collapse/expand all (try expanding an already expanded node etc)
- Making changes to the Storybook (i.e. HMR functionality), does the tree update accordingly and scroll into view?
- What happens if the selected story becomes broken?
- Enable "highlight updates" in React DevTools. Does it rerender an excessive amount of things?

## TO DO

- [x] Link up search results / recently viewed to actually open the selected result
- [x] Properly sync up search input state with Downshift state
- [x] Show the first story when clicking to expand a component for the first time
- [x] Search by and highlight the component path as well
- [ ] Design approval by @domyen 
- [x] Cleanup old stuff
- [ ] QA

ghengeveld committed:
Tweak search input style for recognizability.

ghengeveld committed:
Merge branch 'next' into 10757-better-search

ghengeveld committed:
Tweaks for dark mode compatibility.

ghengeveld committed:
Tweak visual style of Tree items for legibility and contrast.

ghengeveld committed:
Replace Search and Tree in sidebar and implement proper keyboard navigation.

ghengeveld committed:
Fix selecting search result and keyboard navigation.

ghengeveld committed:
Move some stuff around.

ghengeveld committed:
Refactor to custom hooks. Fix docs mode. Cleanup some old stuff.

ghengeveld committed:
Omit RESULTS heading.

ghengeveld committed:
Select the first story when clicking a collapsed component node.

ghengeveld committed:
Using current path to determine new paths is unreliable.

ghengeveld committed:
Change icon color based on theme.

ghengeveld committed:
Color tweaks.

ghengeveld committed:
Match on story path as well.

ghengeveld committed:
Favor name over path matches when searching.

ghengeveld committed:
Move stuff into utils and clean up some old stuff.

ghengeveld committed:
Selected storyId can be null. Pull out lastViewed logic to a custom hook.

ghengeveld committed:
Fix various warnings.

ghengeveld committed:
Fix util tests.

ghengeveld committed:
Fix callback memoization.

ghengeveld committed:
Remove unused variable.

shilman commented:
@ghengeveld this is looking amazing! loving the snappy search and the recently searched list.

```md
small bug: grabs keyboard focus

1. Open a story in `official-storybook` that has controls
2. Start typing in a text field
3. Type `/`

Old behavior: adds "/" to text field
New behavior: adds "/" to text field AND sets focus to search field
```

ghengeveld committed:
Fix compile errors.

ghengeveld committed:
Replace default exports with named exports.

shilman commented on lib/ui/src/components/sidebar/Explorer.stories.tsx:
```diff
@@ -0,0 +1,78 @@
+export default {
+  component: Explorer,
+  title: 'UI/Sidebar/Explorer',
+  excludeStories: /.*Data$/,
```
FWIW, i don't think we really need this anymore with args

shilman commented on lib/ui/src/components/sidebar/Explorer.tsx:
```diff
@@ -0,0 +1,40 @@
+            {...ref}
+            key={refId}
+            isBrowsing={isBrowsing}
+            selectedId={selected && selected.refId === ref.id ? selected.storyId : null}
```
```suggestion
            selectedId={selected?.refId === ref.id ? selected.storyId : null}
```

shilman commented on lib/ui/src/components/sidebar/Explorer.tsx:
```diff
@@ -0,0 +1,40 @@
+            key={refId}
+            isBrowsing={isBrowsing}
+            selectedId={selected && selected.refId === ref.id ? selected.storyId : null}
+            highlightedId={highlighted && highlighted.refId === ref.id ? highlighted.storyId : null}
```
```suggestion
            highlightedId={highlighted?.refId === ref.id ? highlighted.storyId : null}
```

shilman commented on lib/ui/src/components/sidebar/data.test.ts:
```diff
@@ -32,6 +32,7 @@ const a1: Item = {
   isComponent: false,
   isRoot: false,
   parent: 'a',
+  args: [],
```
```suggestion
  args: {},
```

???

shilman commented on lib/ui/src/components/sidebar/data.test.ts:
```diff
@@ -52,6 +53,7 @@ const b1: Item = {
   isRoot: false,
   isComponent: false,
   parent: 'b',
+  args: [],
```
```suggestion
  args: {},
```

shilman commented on lib/ui/src/components/sidebar/data.test.ts:
```diff
@@ -62,6 +64,7 @@ const b2: Item = {
   isRoot: false,
   isComponent: false,
   parent: 'b',
+  args: [],
```
```suggestion
  args: {},
```

shilman commented on lib/ui/src/components/sidebar/data.test.ts:
```diff
@@ -103,6 +106,7 @@ describe('collapse all stories', () => {
         isRoot: false,
         isComponent: true,
         isLeaf: true,
+        args: [],
```
```suggestion
        args: {},
```

shilman commented on lib/ui/src/components/sidebar/data.test.ts:
```diff
@@ -114,6 +118,7 @@ describe('collapse all stories', () => {
         isRoot: false,
         isComponent: true,
         isLeaf: true,
+        args: [],
```
```suggestion
        args: {},
```

shilman commented on lib/ui/src/components/sidebar/useLastViewed.ts:
```diff
@@ -0,0 +1,51 @@
+
+const retrieveLastViewedStoryIds = (): StoryRef[] => {
+  try {
+    const raw = window.localStorage.getItem('lastViewedStoryIds');
```
FYI, we use https://www.npmjs.com/package/store2 in other parts of the code. not sure whether it matters.

ghengeveld replied:
👍

shilman commented on lib/ui/src/components/sidebar/utils.ts:
```diff
@@ -0,0 +1,88 @@
+
+export const createId = (id: string, prefix: string) => `${prefix}_${id}`;
+
+export const get = memoize(1000)((id: string, dataset: Dataset) => dataset[id]);
```
do these memos actually improve perf?

ghengeveld replied:
Not sure. They were there before: https://github.com/storybookjs/storybook/blob/next/lib/ui/src/components/sidebar/Tree/utils.ts#L58

ghengeveld replied:
In the newly added places they are needed to prevent unnecessary rerenders I believe.

shilman reviewed (approved):
Fantastic stuff @ghengeveld 👏 👏 👏

ghengeveld committed:
Update lib/ui/src/components/sidebar/Explorer.tsx

Co-authored-by: Michael Shilman <shilman@users.noreply.github.com>

ghengeveld committed:
Update lib/ui/src/components/sidebar/Explorer.tsx

Co-authored-by: Michael Shilman <shilman@users.noreply.github.com>

ghengeveld committed:
Don't need excludeStories anymore.

ghengeveld committed:
Update lib/ui/src/components/sidebar/data.test.ts

Co-authored-by: Michael Shilman <shilman@users.noreply.github.com>

ghengeveld committed:
Update lib/ui/src/components/sidebar/data.test.ts

Co-authored-by: Michael Shilman <shilman@users.noreply.github.com>

ghengeveld committed:
Update lib/ui/src/components/sidebar/data.test.ts

Co-authored-by: Michael Shilman <shilman@users.noreply.github.com>

ghengeveld committed:
Update lib/ui/src/components/sidebar/data.test.ts

Co-authored-by: Michael Shilman <shilman@users.noreply.github.com>

ghengeveld committed:
Update lib/ui/src/components/sidebar/data.test.ts

Co-authored-by: Michael Shilman <shilman@users.noreply.github.com>

ghengeveld committed:
Use store2 instead of direct localStorage.

domyen commented:
Nice work Gert! 

It seems like the CI isn't working for some reason. That's preventing me from seeing the published SB and reviewing the UI. I think `next` may be working again.

ghengeveld committed:
Allow clearing the last viewed history.

shilman committed:
Fix typings

domyen commented:
Well done 👏! This is impressive. 

**Requests:** 
- Make sidebar wider from 210px → 230px. The width was set at 210px in 5.0. With a year of real use, that width is too tight given the length of many story names. We never got around to changing it.
- The keyboard shortcuts in the (...) menu should be styled like the [/] in the search field. [i.e., like this](https://projects.invisionapp.com/d/main#/console/20491930/430355148/preview)

**Search UI/UX:**
- Show more/all results in search. Showing so few results makes the search _seem_ underpowered (which it's not). 
- Change "Show all x results" to "Show x more results". When I click on that button I want to know how many _more_ results will be revealed. I care less about the total number of results.
- In the search results and recently opened, the `›` makes the path look like a breadcrumb when its not. Change to `/`. This aligns with how nested file structures are displayed.
- Change search cancel button right position to `right: 8px`
- Preserve search input value, when I click outside the search field it shouldn't clear the search input. When I'm in the search modality I want search results to persist until I manually clear it with the (X) button. This allows me to jump between search results to find the component I want.
- Overflows for path should also cut off at the end of the path with an (...) ![image](https://user-images.githubusercontent.com/263385/95226620-2971a000-07cb-11eb-9555-eaae8b349d63.png)
- Add the full path in the `title` attribute so the user can see the path on hover. e.g. `Storybook Design System / Design system / Avatar`


p.s., The "clear history" button is a nice detail. 👨‍🍳 💋

tooppaaa commented on lib/ui/src/components/sidebar/Search.tsx:
```diff
@@ -1,145 +1,320 @@
-}));
+  const selectStory = useCallback(
+    (id: string, refId: string) => {
+      if (api) api.selectStory(id, undefined, { ref: refId !== DEFAULT_REF_ID && refId });
```
```suggestion
     api?.selectStory(id, undefined, { ref: refId !== DEFAULT_REF_ID && refId });
```

tooppaaa commented on lib/ui/src/components/sidebar/Search.tsx:
```diff
@@ -1,145 +1,320 @@
+    const focusSearch = (event: KeyboardEvent) => {
+      if (isLoading || !inputRef.current) return;
+      if (event.shiftKey || event.metaKey || event.ctrlKey || event.altKey) return;
+      if (event.key === '/' && inputRef.current !== document.activeElement) {
```
There's a disable shortcut option that is not taken into account here, should it be ? (and hide shortcut in the meantime)

ghengeveld replied:
Fixed. Good call!

tooppaaa reviewed (commented):
Lots of good stuff !  Hard to review everything, I might do another round sometime soon :)

ghengeveld committed:
Change default sidebar width to 230px (visually).

ghengeveld committed:
Show 50 search results by default.

ghengeveld committed:
'Show X more results' instead of 'Show all X results'.

ghengeveld committed:
Show 20 results by default, so the 'show more' button is visible on most desktop displays.

ghengeveld committed:
Change path separator to '/'.

ghengeveld committed:
Fix clear button alignment.

ghengeveld committed:
Add title attribute to reveal full path.

ghengeveld committed:
Styling for keyboard shortcuts in menu.

ghengeveld commented:
> * Make sidebar wider from 210px → 230px. The width was set at 210px in 5.0. With a year of real use, that width is too tight given the length of many story names. We never got around to changing it.

I've implemented this as the default width now, but kept the minimum width at 210px. Let me know if you want it to be a hard minimum.

> The keyboard shortcuts in the (...) menu should be styled like the [/] in the search field. i.e., like this

Missed that. Done now.

> Show more or even all results in search. Showing so few results makes the search seem underpowered (which it's not).

Changed it to show 20 results by default. I picked this number because it means the "Show more" button is visible on most desktop displays, but only at the very bottom. This is useful to gauge the total number of results without having to scroll.

> Change "Show all x results" to "Show x more results".

Good call 👍  Fixed it now.

> In the search results and recently opened, the `›` makes the path look like a breadcrumb when its not. Change to `/`. This aligns with how nested file structures are displayed.

Even though visually I like the `›` more, I agree this is better for consistency / familiarity 👍

> Change search cancel button right position to right: 8px

I had this aligned with the `[/]` icon, but you're right it looks better when moved slightly further to the right.

> Preserve search input value, when I click outside the search field it shouldn't clear the search input. When I'm in the search modality I want search results to persist until I manually clear it with the (X) button. This allows me to jump between search results to find the component I want.

Sounds like a good idea but it has some implications (both technically and functionally). Right now "recently viewed" is implemented like they're default search results, so this behavior would also apply there (unless I hack around it). Then there's the issue of focus management and keyboard listeners. When you select a result, we open the component which may claim focus and/or register keyboard listeners. That means we can lose focus on the search field and/or the ability to close the search by pressing `Escape`. I suggest we gather more feedback before going down this road.

> Overflows for path should also cut off at the end of the path with an (...)

It does, but in this case the "Design System" part is simply too short to ever get cut off because it's no longer than 50% of the total width. You can verify by inspecting the element and making the string longer.

> Add the full path in the title attribute so the user can see the path on hover. e.g. Storybook Design System / Design system / Avatar

Good call 👍 Added.

ghengeveld committed:
Update lib/ui/src/components/sidebar/Search.tsx

Co-authored-by: Clément DUNGLER <clementdungler@gmail.com>

ghengeveld committed:
Respect the enableShortcuts setting for focussing the search input.

ghengeveld committed:
Merge branch 'next' into 10757-better-search

shilman commented:
@ghengeveld it looks like the new DOM structure is breaking our cypress e2e tests:

```
  1) Navigation
       should search navigation item:
     AssertionError: Timed out retrying: Expected to find element: `#storybook-explorer-searchfield`, but never found it.
      at Context.eval (http://localhost:8001/__cypress/tests?p=cypress/integration/navigation.spec.ts:162:12)

  2) Navigation
       should display no results after searching a non-existing navigation item:
     AssertionError: Timed out retrying: Expected to find element: `#storybook-explorer-searchfield`, but never found it.
      at Context.eval (http://localhost:8001/__cypress/tests?p=cypress/integration/navigation.spec.ts:168:12)

  3) Routing
       should navigate to story addons-a11y-basebutton--default:
     AssertionError: Timed out retrying: Expected to find element: `#addons-a11y-basebutton--label`, but never found it.
      at Context.eval (http://localhost:8001/__cypress/tests?p=cypress/integration/navigation.spec.ts:175:12)
```

ghengeveld committed:
Revert "Update lib/ui/src/components/sidebar/Search.tsx"

This reverts commit ea9a9a84e556749c9c48373d2a1f1c5464079138.

ghengeveld committed:
Make sure leaf nodes always have args.

ghengeveld committed:
Add missing id attributes.

domyen commented:
### **How many search results**
> Changed it to show 20 results by default. I picked this number because it means the "Show more" button is visible on most desktop displays, but only at the very bottom. This is useful to gauge the total number of results without having to scroll.

On further thought, I suggest removing the show more altogether (unless there are _significant_ performance benefits). This makes it easier to skim all the results without leaving the search field.

<details>
<summary>More background</summary>

This isn't a typical search where result count is useful for things like pagination or to get a sense of information scale. The pattern here is find/filter. Folks are filtering by changing the query – similar to VSCode `cmd+t` and skimming the results. They're more likely to keep the search field focused and guess different queries than click out to paginate the results. That leads me to suggest showing all the results at once to let users skim the full list instead of making them click more.

iirc, Yann mentioned this behavior while we were researching. Something to the effect of "I only care about the first few results and will keep typing until I see what I want".
</details>

### **Preserve search value**
> Right now "recently viewed" is implemented like they're default search results, so this behavior would also apply there (unless I hack around it). Then there's the issue of focus management and keyboard listeners. When you select a result, we open the component which may claim focus and/or register keyboard listeners. That means we can lose focus on the search field and/or the ability to close the search by pressing Escape. I suggest we gather more feedback before going down this road.

I don't think punting on this workflow makes sense. The current UX is intended for the ideal case. The user knows exactly what they want to click on. But in unideal cases, the flow becomes cumbersome. 

Check out this real-life scenario from Chromatic's Storybook:
1. You're working on a `PullRequests` feature
2. Search for `PullRequests` to get a list of components below (there are at least 12 results for this term)
![image](https://user-images.githubusercontent.com/263385/95357616-124bb480-0896-11eb-9679-b3797c90af0a.png)
3. Click on one to see it 
4. Realize that it's not the component you want 🤦‍♀️
5. Now what?

**Option: Leave behavior as is**
Folks will have to type out `PullRequests` again, then click on another result. They'll type `PullRequests` again and again to see the results because clicking on one will clear the search field. 

**Option: Change behavior to preserve search results**
Folks will click again on the search field whose value remains `PullRequests` to see the results. They click the next result to view the component. If that's not the component they're looking for, they can click on the search field again to view the results.

ghengeveld committed:
Rename various props to clarify the distinction between storyId, itemId and the id with refId.

ghengeveld committed:
Better logo alignment.

ghengeveld committed:
Update e2e test for search.

ghengeveld committed:
Try to fix e2e test.

ghengeveld committed:
Prevent hijacking keyboard event listeners in unrelated sections of the UI.

ghengeveld committed:
Make sure Cypress can know when a story is selected.

ghengeveld committed:
Better story layout.

ghengeveld committed:
Ensure parent is expanded before trying to click an item.

ghengeveld committed:
Enhance contrast.

ghengeveld committed:
Try with the 'attribute' method.

ghengeveld committed:
Try with the 'attr' method.

ghengeveld committed:
Fix clear icon color.

ghengeveld committed:
More contrast, especially for the dark theme.
